### ISSUE TITLE
[ BUG REPORT ] - High CPU Usage When Not Idling

### ISSUE THREAD

**Issue by zeropointnothing** [Just-Natsuki-Team/NatsukiModDev#887]
# Overview:
When Natsuki is not idling or no topic menu/topic is active, the program seems to use an unusual amount of CPU on one core, even when the window is minimized.

I have noticed this issue persistently across computers, both on Windows and Arch Linux.

## Replication steps:
- Open JN

## Additional information:
Here's some info on both of my systems.

### Windows

OS: Windows 11, version 10.0.22631 Build 22631
CPU: AMD Ryzen 7 5800
GPU: NVIDIA GeForce RTX 3060
JN: v1.3.4

### Linux

OS: Arch Linux
Kernel Version: 6.7.9-arch1-1
DE: KDE Plasma 6 (Wayland and Xorg)
WM: kwin
CPU: AMD Ryzen 7 7840U
GPU AMD ATI c1:00.0 Phoenix1
JN: v1.3.4

Image from btop showing CPU usage. All of it seems to be on one core, but it stills gives my poor laptop a workout.
![image](https://github.com/Just-Natsuki-Team/NatsukiModDev/assets/116678675/3713125c-8fc1-4de7-8b0f-31f3fd9a9c18)

**Comment by Blizzardsev** [comment]
Hi @zeropointnothing - just to clarify: do you mean to say this is an issue when Natsuki is going through her stand-by animations (looking left/right, smiling, etc.) specifically?

Also, can you confirm if you have Weather set to Real Time and if this makes any difference or none at all?

**Comment by zeropointnothing** [comment]
Yes, after Natsuki enters one of her idles, or I interact with her with the "Talk" button, the CPU usage returns to normal. Only when she's in stand-by does the problem appear. Also, none of the weather options seem to change anything.

**Comment by RZed01** [comment]
I believe this problem could also be one of the main reasons for the lower frame rates that occur in her default (normal/standby) state, previously reported in #855 and #824. It could be an indication that Ren'Py is a single threaded process, which would explain this.

**Comment by Blizzardsev** [comment]
We'll continue to investigate this as replication is proving difficult - might be more of an issue with Ren'Py and memory handling over longer play sessions.

**Comment by Blizzardsev** [comment]
Hi @zeropointnothing - thank you for your patience, we've identified the issue and have a solution.

In short, the problem lay with how we handled idle sprites for Natsuki. The current release (and all prior releases) use a Ren'Py function that automatically updates the idle image based on a list of conditions when that image is displayed.

Given the amount of idle variations we have and the function itself not implementing any form of caching, it meant excess CPU time was being spent by the game constantly checking the conditions and updating the image whenever the game idled, even if nothing had changed in those conditions mentioned earlier. This explains the performance difference you saw between idling and when in a menu, as Nat switched sprites and the game no longer ran the updates.

Our solution will ensure the game only updates her sprite when necessary, and should result in increased performance during idle periods: we can already see a big improvement using Performance Monitor.

We'll include this fix in the next patch.
